### ISSUE TITLE
Allow to explicitly pass parameter types via JSDoc

### ISSUE THREAD

**Issue by ifeltsweet** [Microsoft/TypeScript#27387]
## Search Terms

jsdoc generics constraints

## Suggestion

It seems like it is not possible via JSDoc to explicitly tell compiler which type parameters to pass to a generic function.

## Use Cases

In TypeScript it is possible to explicitly pass type parameters such as `mongoose.model<Model, Schema>('modelName', Schema)` while I could not find a way to do same with JSDoc.

## Examples

`mongoose.model` has two signatures, first one with one type parameter and the other with two. To make use of the second signature we must pass types explicitly.

```ts
export function model<T extends Document>(name: string, schema?: Schema, collection?: string,
    skipInit?: boolean): Model<T>;
  export function model<T extends Document, U extends Model<T>>(
    name: string,
    schema?: Schema,
    collection?: string,
    skipInit?: boolean
  ): U;
```
```ts
//  Choose second signature in typescript.
const model = mongoose.model<Model, Schema>('modelName', Schema);

// With JSDoc, compiler always chooses first signature and we receive a type mismatch error.
/** @type {Schema & mongoose.Model<Model>} */
const model = mongoose.model('modelName', Schema);

// But something like this would be great.
const model = mongoose.model/**<Model, Schema>*/('modelName', Schema);
```

My apologies if this is already possible, but I've spend almost a week battling this.

Related: https://github.com/Microsoft/TypeScript-Node-Starter/issues/101

## Checklist

My suggestion meets these guidelines:
* [?] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [?] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [?] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by Stuk** [comment]
I would also like to see something like this. Currently it's really hard to use functions with generic types from JS.

**Comment by ekulabuhov** [comment]
Another use case with React Hooks:

```js
// useState<string> is derived correctly
const [aString, setAString] = useState("default value");
```

Removing default value there is no way to pass type info:
```js
// useState<any>
const [aString, setAString] = useState();
```

**Comment by pm0u** [comment]
Anything on this? Also stuck on the useState example as shared above:

```
// Doesn't set type properly
const [error, setError] = /** @type {React.useState<?string>} */ useState(null);
// Nor does this
  /** @type {[?string, React.Dispatch<React.SetStateAction<?string>>]} */
  const [error, setError] = /** @type {React.useState<?string>} */ useState(
    null
  );


// we have an error we want to set now.
setError('Error!') // Shows a type error
```

**Comment by robertknight** [comment]
For functions that take an argument of the generic type, casting the argument itself is reasonably practical. We've settled on these patterns with `useState` and `useRef` in React for example:

```js
const [thing, setThing] = useState(/** @type {SomeType|null} */ (null));
const widgetRef = useRef(/** @type {HTMLElement|null} */(null));
```

In cases where there is no argument for everything else to be inferred from, it gets much more verbose unfortunately. What would be helpful is being able to do something like:

```js
const aSet = /** @type {?<string>} */(new Set());
```

Where the `?` is inferred as `Set`. In this example it only saves a few characters, but it could be a lot more if the generic type name is lengthy and/or has to be imported from somewhere.

**Comment by k-yle** [comment]
> Where the `?` is inferred as `Set`. In this example it only saves a few characters, but it could be a lot more if the generic type name is lengthy and/or has to be imported from somewhere.

Something like that would be awesome, it would majorly simplify the code required for `React.forwardRef` and `React.memo`:

```diff
- const Component = /** @type {React.ForwardRefExoticComponent<React.RefAttributes<Props>>} */ (React.forwardRef(() => { ... }))
+ const Component = /** @type {?<Props>} */ (React.forwardRef(() => { ... }))

**Comment by trusktr** [comment]
Calling functions with generic types is important in AssemblyScript. This would help to possibly get us to a point where we can write working plain JS, but also compile the same code to WebAssembly.

**Comment by sdegutis** [comment]
Casting the argument sometimes is too verbose. For example, when you pass a map of strings to T, now you have to cast the whole map to `/** @type {{ [key: string]: MyInterface }} */`. When you want the map values to be `() => T` or even more complex, the cast gets uglier.

I'd suggest changing the parser to look for a `/** @typeparam {MyInterface} */` in between the function expression and the function argument-set parentheses.

**Comment by petrkrejcik** [comment]
@weswigham Hi, I would like to kindly as if the discussion has evolved? (you've added a tag "In Discussion")
Could we hope for this feature?

**Comment by phil294** [comment]
> JSDoc comments are for documentation. We should avoid repurposing JSDoc comments for features that aren't documentation

@trusktr JSDoc is has not been only for documentation for a long time now. It supports [almost anything TS does](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html), with [very few exceptions](https://github.com/microsoft/TypeScript/issues/30624). As for syntax, the `/** */` is well-established and parsed by the compiler if you're programming in JS.

---

> ```ts
> // But something like this would be great.
> const model = mongoose.model/**<Model, Schema>*/('modelName', Schema);
> ```

There is already existing JSDoc syntax for *declaring* generic types: [`@template`](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template). So surely this tag should be reused (?)
```ts
const model = mongoose.model/** @template Model, Schema */('modelName', Schema);
```

**Comment by skapxd** [comment]
I use this way to send types as parameters, but it's a bit dirty

```javascript
/**
 * @template T
 * @param {Object} props
 * @param {string} props.key
 * @param {T} [props.templateType]
 * @returns {T}
 */
const getLocalStorage = ({ key = "" }) => {
  const data = localStorage.getItem(key);
  return JSON.parse(data);
};

/**
 * @type {{
 * id: number,
 * name: string,
 * phoneNumber: number
 * }}
 */
const templateType = null;
getLocalStorage2({
  templateType,
  key: "somebody",
});


```

**Comment by wenq1** [comment]
is there any plan moving this forward? @RyanCavanaugh

**Comment by RedGuy12** [comment]
Any updates?

**Comment by phaux** [comment]
BTW I just found out this works:

```js
import { useState } from "react"

/** @type {typeof useState<number>} */
const useNumberState = useState

const [value] = useNumberState(123)
```

or

```js
import { useState } from "react"

const [value] = /** @type {typeof useState<number>} */ (useState)(123)
```

equivalent to

```ts
import { useState } from "react"

const [value] = useState<number>(123)
```

**Comment by noinkling** [comment]
@phaux Yes, they're called "instantiation expressions" (see #47607) and were added in 4.7

Another option is doing something like:

/** @type {ReturnType<typeof fn<SomeType>>} */
const myVar = fn('whatever')

No intermediate variables or nasty inline casting syntax, but I guess less semantically 'pure', and can't be used in certain edge cases where `ReturnType` doesn't work well.

**Comment by zhenzhenChange** [comment]
I don't think there's much of a problem with using type assertions (type conversions) :

```js
const elRef = useRef(/** @type {HTMLElement | null} */ (null));
const [value, setValue] = useState(/** @type {SomeType | null} */ (null));
```

Only the readability of the code decreases with the increase of type comment.

---

[TypeScript Doc](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#casts)

![image](https://user-images.githubusercontent.com/51586637/188396991-bc431286-85ab-4174-9288-d88a05b71808.png)

[Cast Syntax](https://database.guide/how-cast-works-in-sql-server/)

![image](https://user-images.githubusercontent.com/51586637/188397108-c62e7997-0035-478d-9441-4e670d39dd9f.png)

**Comment by milahu** [comment]
https://github.com/microsoft/TypeScript/issues/27387#issuecomment-659671940

this should be in https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#template

**Comment by Jemt** [comment]
+1

**Comment by thepassle** [comment]
As mentioned in #56102 , I think it'd be really great to add better support for this. Lots of projects are using JSDoc for typechecking instead of full-on TS (either approaches are valid and fine, lets not start that discussion), and it seems to me like better support for generics in jsdoc is long overdue.

I also think the example proposed by the OP would be a great candidate:
```js
const model = mongoose.model/**<Model, Schema>*/('modelName', Schema);
// or perhaps some variation of an @ tag, that seems more in-line with the current approach JSDoc takes
const model = mongoose.model/** @generic {<Model, Schema>} */('modelName', Schema);
```
What I think is nice about these approaches is that:
- Still valid JS, so will run without a compilation/build step
- Visually similar to the TS way of typing this

Would love to discuss and see what we can do to get the ball rolling on this, especially considering that this issue has been open for 5 years :)

**Comment by noinkling** [comment]
@thepassle Technically this is already solved as per https://github.com/microsoft/TypeScript/issues/27387#issuecomment-1223795056

For example you can now do:

```
const model = /** @type {typeof mongoose.model<Model, Schema>} */(mongoose.model)('modelName', Schema);
```

If you're aware of this but your argument is that you would prefer for there to be a more ergonomic syntax, probably best to say that explicitly.

**Comment by mhofman** [comment]
@noinkling, those are not equivalent. A generic method would become unbound from its target if aliased or type annotated that way.

**Comment by noinkling** [comment]
> @noinkling, those are not equivalent. A generic method would become unbound from its target if aliased or type annotated that way.

You mean methods won't have the right `this` value, essentially? I can see the issue with the alias approach in those cases, but nothing new there. Can you give an example where the casting syntax could be problematic/have a different result compared to TS syntax?

**Comment by gustavopch** [comment]
```ts
const model = /** @type {typeof mongoose.model<Model, Schema>} */(mongoose.model)('modelName', Schema)
```
is equivalent to:
```ts
const model = (mongoose.model as typeof mongoose.model<Model, Schema>)('modelName', Schema)
```
not to:
```ts
const model = mongoose.model<Model, Schema>('modelName', Schema)
```
The fact that this ugly but useful workaround exists doesn't mean a proper syntax for actually passing type parameters (instead of casting the whole function) shouldn't be pursued.

**Comment by noinkling** [comment]
@gustavopch It didn't exist when this issue was created was the point, it was impossible to provide type params unless they corresponded to runtime params. In that sense the person who created the issue might consider it "solved" now. Personally I've learned to take what I can get when it comes to JSDoc, given that it's something of a second-class citizen for TS. But you're right, it would be nice to have a nicer syntax, just wanted to make it clear that there's an option that works now (even if it isn't pretty).

**Comment by phaux** [comment]
I return to this thread to point out that the "solution" with instantiation expressions doesn't seem to work when the function has multiple overloads:

```js
const fileInput =
  /** @type {typeof document.querySelector<HTMLInputElement>} */
  (document.querySelector)("input[type=file]")
```

It seems to check all the overloads at once instead of choosing only one:

```
Type 'HTMLInputElement' does not satisfy the constraint 'keyof HTMLElementDeprecatedTagNameMap'. deno-ts(2344)
Type 'HTMLInputElement' does not satisfy the constraint 'keyof HTMLElementTagNameMap'. deno-ts(2344)
Type 'HTMLInputElement' does not satisfy the constraint 'keyof MathMLElementTagNameMap'. deno-ts(2344)
Type 'HTMLInputElement' does not satisfy the constraint 'keyof SVGElementTagNameMap'. deno-ts(2344)
```

In TypeScript it works:

```ts
const fileInput =
  document.querySelector<HTMLInputElement>("input[type=file]")
```

**Comment by noinkling** [comment]
@phaux You're right, worth noting it could potentially be considered a bug: #51694

**Comment by s3xysteak** [comment]
[playground](https://www.typescriptlang.org/play/?filetype=js#code/PQKhCgAIUgBAXAngBwKYBNUDNIG8D2ARgFaoDG8AvpABKoBO+UMsyja9SkAFsngHYBXALaEGlZnDb4OXfgENhqPAGd49AJb8A5hOjBw4UDEkJUw5ABt58ZQBVTyefUV471IsUjMDWQfwoNfH5IJ3oVVAB5EgAKTwBKPChIFPpUeEF6EM9wCXAyYLVIT0gAXiSUlN4ALkgARgAaZJSFJVqAciV8AHd23MMC-iKAN3lLAEEy0OcI6OI4knjDYzgkNDwAJXTM-jsUVAAeNdR8HDDZkgO6RgA+G+oQA0GRsYAhKfOo2ITwIA)

That looks good.

```js
/** @type {ReturnType<typeof func<MyType>>} */
const val = func(obj)
```

**Comment by apendua** [comment]
@sandersn I am looking into implementing something like:
```
const model =
  /** @specialize {Model, Schema} */
  mongoose.model('modelName', Schema)
```
Do you think there's a chance this approach will be accepted?

**Comment by sandersn** [comment]
Our approach to JSDoc has evolved since this issue was created. At that time I didn't want to add anything that didn't have precedent in the jsdoc documentation generator, or at least in Closure compiler. However, Typescript is the largest processor of JSDoc at this point, and TS users who want typing are probably the most prolific authors of it. So we've decided on a few new principles:

1. If you're exposing existing TS syntax and semantics to JS, the semantics should be identical and the syntax should be as close as possible. `@import` is a good example, as is your own `@overload` work.
2. However, new tags should be normal tags as much as possible. That means:
   a. Starts with `@alphanumeric`.
   b. Isn't nested -- starts at `@` and continues till the next `@`.
   c. Is used in a `/**` comment directly above a declaration. Both `@import` and `@overload` violate this, for good reasons.

(1) means that TS users will be able to use the new JS syntax with a minimum of surprise. (2) means that jsdoc processors (including TS itself!) will be able to understand the new tag with a minimum of new parsing.

Re (2c), the rough descending order of goodness, or familiarity, in my opinion is:

1. In a `/**` comment directly preceding a declaration -- like `@type`.
2. A `/**` comment not connected to any JS syntax (which is to say, preceding some non-declaration) -- like `@import`.
3. A `/**` comment preceding an expression -- like `@type` as a cast.
4. A `/**` comment preceding another `/**` comment that eventually precedes a declaration -- like `@overload`.
5. A `/**` comment preceding some other, non-top-level node in the parse tree.
6. A `/*::` comment or some other new syntax.
7. A `// @ts-alphanumeric` comment preceding a declaration.

I think we can come up with something for (3).

This is a very long way to say that, yes, we'd consider a tag for this that looks a lot like your `@specialise`. You're probably going to need parentheses around the call expression to make it attach correctly, though.
Also I'm not sure whether `<Model, Schema>` would be better than `{Model, Schema}`. I think it's down to how hard it is to convert *all* the type-parameter syntax to that form. The grammar isn't huge there so probably pretty easy.

@DanielRosenwasser you've thought about jsdoc a lot too *and* you may have interesting insights from working on the types-in-JS proposal, which ran into exactly this problem.

**Comment by apendua** [comment]
I've got a PR ready for the following syntax:
```javascript
let a = /** @specialize <number> */(f());

/** @specialize <string, number> */
let b = f();

/** @specialize {string} */
let c = f`template with ${variable}`;

let d = /** @specialize <string> */(new D());
```
So basically one can put the `@specialize` tag at any parent node that accepts JSDoc, e.g. assignment statement or parethesized expression. We can also choose between wrapping type arguments with `<...>` or `{...}`.

This will also work for JSX, though the element with type arguments will need to put inside curly braces so that we can also add JSDoc there:
```javascript
function MyForm() {
  return (
    <form>
     {/** @specialize {number} */(
       <Input value={0} />
     )}
    </form>
  )
}
```

**Comment by DanielRosenwasser** [comment]
It's something we'll have to consider. It's not perfect, but the solution right now is to cast to a value that's already specialized:

```js
/**
 * @template T
 * @param x {T}
 */
function f(x) {
    return x;
}

let obj = {
    /**
     * @template T
     * @param x {T}
     */
    f(x) {
        return x;
    }
};

let g = /** @type {typeof f<string>} */ (obj.f);
//  ^?
// typeof f<string>

let objG = (/** @type {typeof obj.f<string>} */ (obj.f)).bind(obj);
//  ^?
// typeof obj.f<string>
```

https://www.typescriptlang.org/play/?filetype=js#code/PQKhFgCgAIWgBALgUwLYAcA2BDF0AqUsC62ATtqtAB7QDe+AvkSMFAGYCuAdgMaIBLAPbdo7ABTUAlPSLR5ZZIk5lR1ANxRmkKJiXQhAIwBW0ALyyY86KAhXrcJGiy5kBOQ5LlKNekw-yrAES0pbW4dCKyqo0mvbQ2oxxuvoA5uY2YAiIAJ7obnS5+ULsYgA8AM6IZALcqQB8jLDA0OJGxgB07FJxwC3QAHoA-FApiAYmAOIZ4rbZeQVFyCUTneyV1bUNTayt7V1SUh2GtQAmbSY9UH3yw0A

**Comment by apendua** [comment]
@DanielRosenwasser 

> Though noted above in https://github.com/microsoft/TypeScript/issues/27387#issuecomment-1925609832, it doesn't work in the case of overloads.

Good point! Btw, the reason this currently does not work:
```

**Comment by apendua** [comment]
@DanielRosenwasser 

> Though noted above in https://github.com/microsoft/TypeScript/issues/27387#issuecomment-1925609832, it doesn't work in the case of overloads.

Good point! Btw, the reason this currently does not work:
```javascript
const fileInput =
  /** @type {typeof document.querySelector<HTMLInputElement>} */
  (document.querySelector)("input[type=file]")
 ```
is already pointed out in the other issue, here:

https://github.com/microsoft/TypeScript/issues/51694
 
Do you think there's a chance the other PR will be merged? i.e. https://github.com/microsoft/TypeScript/pull/51695

**Comment by tif-calin** [comment]
I'm a big TypeScript fan that's trying out a new project where I use JSDocs instead of TS. Here are some of the learnings I've been forced to do along the way.

### Passing in a generic
The key here is utilizing `ReturnType`.

TypeScript version:
```ts
const patchReq = async <T>(path: string, data: Partial<T>): T => makeRequest('PATCH', path, data);

const updateBallot = async (ballot: Partial<Ballot>) => patchReq<Ballot>(`elections/${ballot.id}`, ballot);
//      ^? returns Promise<Ballot>
```

JSDocs version:
```js
/** @template T; @param {string} path; @param {Partial<T>} data; @returns {Promise<T>} */
const patchReq = async (path, data) => makeRequest('PATCH', path, data);

/** @param {Partial<Ballot>} ballot; @returns {ReturnType<typeof patchReq<Ballot>>} */
const updateBallot = async (ballot) => patchReq(`ballots/${ballot.id}`, ballot);
```

### React useState
This was already mentioned above, but I think it's useful to rewrite. Typecasting the initial value has never failed me. Make sure to remeber the parentheses 
```js
const [optionalTitle, setOptionalTitle] = useState(/** @type {null | string} */(null));
//      ^? null | string
```
